### ISSUE TITLE
Conditional metrics ?

### ISSUE THREAD

**Issue by sdesrozis** [pytorch/ignite#816]
## ❓ Questions/Help/Support

How do you deal the confusion matrix metric ? It is an expensive metric that is interesting to have punctually during the training. A naive way is to attach `ConfusionMatrix` to evaluators but it could have a performance issue. Otherwise, a dedicated evaluator could be created but it multiplies evaluators... Any advice on that point ? My feeling is we need "conditional" metrics triggered wrt to specific predicat.

**Comment by vfdev-5** [comment]
@sdesrozis I think this depends on application and the additional cost needed to be quantified. 
You can take a look at this example: https://github.com/pytorch/ignite/blob/master/examples/references/segmentation/pascal_voc2012/code/scripts/common_training.py#L137
Where validation IoU metric is computed with CM and evaluation is triggered every n epochs.
HTH

**Comment by sdesrozis** [comment]
Ok, thank you for your answer :+1:
I noticed that because I did a classification training (approx 20 classes) and I had to move CM to postprocess due to performance. Tho, it wasn't a blocking issue.
In your example, how would you add additionnal metrics running at each epoch completed ? I don't see any other solution than adding an evaluator.

**Comment by vfdev-5** [comment]
> In your example, how would you add additionnal metrics running at each epoch completed ? I don't see any other solution than adding an evaluator.

Yes, actually, there is no simple solution without adding another evaluator.
We can think of this as a feature request :)

**Comment by sdesrozis** [comment]
It would be a pleasure for me to contribute on that point 😊

**Comment by vfdev-5** [comment]
That's good ! So, now we need to decide on the API for such feature. Something simple and looking cool :)
For example, how about that:
```python
trainer = ...
evaluator = ...
metrics = {
   "Accuracy": Accuracy(),
   "CM": ConfusionMatrix(num_classes=20)
}

@trainer.on(Events.EPOCH_COMPLETED)
def run_fast_evaluation(_):
    metrics["CM"].disable()  # metric CM is not computed
    evaluator.run(test_dataloader)


@trainer.on(Events.EPOCH_COMPLETED(every=5))
def run_extensive_evaluation(_):
    metrics["CM"].enable()
    evaluator.run(test_dataloader)

...

trainer.run(train_dataloader, max_epochs=100)
```

Obviously, with such solution, when logging the values from `evaluator.state.metrics` on every evaluator's `Events.COMPLETED`, accuracy will be updated every training epoch, but CM will be updated by steps, like that
```
^ CM 
|                    * * * * *
|          * * * * *
| * * * * *
|-------------------------------> epochs
```
which is not that good. To handle, that, probably, `metrics["CM"].disable()` should remove the key from `evaluator.state.metrics`, what can it be dangerous (?).

Thougths ?
cc @justusschock @ykumards

**Comment by sdesrozis** [comment]
If CM is bot computed, dict metrics shouldn’t contain key « CM ». So easily checkable for logging. My question is should we disable explicitly metric or should we apply a predicat or filter to enable / disable implicitly during event handling.

**Comment by vfdev-5** [comment]
> If CM is bot computed, dict metrics shouldn’t contain key « CM »
Correct, the state will be reset on each run.

Metrics are "attached" (they add multiple handlers), so, currently, there is no simple API to enable/disable a metric. I agree with you that we can also propose a sort of `detach` method that can remove all previously attached handlers. I maybe this is a better API than previously proposed `disable`/`enable`.

In my previous idea, it was thought as internal redirection of `iteration_completed` and `completed` methods. It is more simple to implement etc, but maybe it gives a not that clear API.

So, let's think about the following API:
```python
trainer = ...
evaluator = ...
metrics = {
   "Accuracy": Accuracy(),
   "CM": ConfusionMatrix(num_classes=20)
}

 for name, metric in metrics.items():
        metric.attach(evaluator, name)


@trainer.on(Events.EPOCH_COMPLETED)
def run_fast_evaluation(_):
    metrics["CM"].detach()  # metric CM is completely detached from evaluator
    evaluator.run(test_dataloader)


@trainer.on(Events.EPOCH_COMPLETED(every=5))
def run_extensive_evaluation(_):
    metrics["CM"].attach(evaluator, "CM")
    evaluator.run(test_dataloader)

...

trainer.run(train_dataloader, max_epochs=100)
```

**Comment by sdesrozis** [comment]
I like this API which is simple to deploy. An alternative is to try to have only one call to evaluator.run() (cheaper, less data load) and use the filter (every=5) attached to the metric itself. It shouldn’t solve logging issues (we have to know what is computed). Sorry I can’t propose snippets atm because my only access to github is my smartphone. Tho, I don’t know if it’s feasible neither user friendly...

**Comment by sdesrozis** [comment]
Nb : `detach` should have an argument ˋevaluator` because it shouldn’t be a collective detachment, should it?

Maybe the condition ˋtrainer.state.epoch % 5 == 0` could be managed as an extra argument of the first attachment loop (only for CM metric). I try to illustrate tomorrow (I’m in France).

**Comment by vfdev-5** [comment]
> Nb : detach should have an argument evaluator because it shouldn’t be a collective detachment, should it?

Technically with the current implementation, yes. But I think now we need to go futher with `Metric` class and register as a weakref the engine. As we'll allow attach/detach we need to make sure to handle properly multiple faulty attaches etc. So, once metric know its engine, we wont need to insert engine as argument to `detach`.

> Maybe the condition trainer.state.epoch % 5 == 0 could be managed as an extra argument of the first attachment loop (only for CM metric). I try to illustrate tomorrow (I’m in France)

I think this is tricky to do as `attach` uses evaluator, but event filtering is on trainer. But anyway if you have a clear idea how to do it, why not. Actually, me too in France, oui, il faut aller dormir :)

> I just investigate an option, hope it’s ok for you.

I'm OK with that.

**Comment by sdesrozis** [comment]
> Technically with the current implementation, yes. But I think now we need to go futher with `Metric` class and register as a weakref the engine. As we'll allow attach/detach we need to make sure to handle properly multiple faulty attaches etc. So, once metric knows its engine, we wont need to insert engine as argument to `detach`.
I understand that the metric could encapsulate the engine. But a metric could be related with several engines. Maybe I didn’t catch your idea.

> I think this is tricky to do as `attach` uses evaluator, but event filtering is on trainer.
It looks like decoration pattern in your implementation. 

> Actually, me too in France, oui, il faut aller dormir :)
Super :)

**Comment by sdesrozis** [comment]
I implemented the following test
```python
class DummyMetric(Metric):
    def __init__(self, name, output_transform=lambda x: x):
        self._name = name
        self.call_count_reset = 0
        self.call_count_compute = 0
        self.call_count_update = 0
        super(DummyMetric, self).__init__(output_transform=output_transform)

    def reset(self):
        print(f"{self._name} {id(self)} : reset")
        self.call_count_reset += 1

    def compute(self):
        print(f"{self._name} {id(self)} : compute")
        self.call_count_compute += 1

    def update(self, output):
        print(f"{self._name} {id(self)} : update")
        self.call_count_update += 1


def test_metric_filtering():

    trainer = Engine(lambda e, b: None)
    evaluator = Engine(lambda e, b: None)

    # metrics creation
    # Remark: reset() is called at creation
    dummy1 = DummyMetric('dummy1')
    dummy2 = DummyMetric('dummy2')

    # dummy2 is attached
    dummy2.attach(evaluator, 'dummy2')

    def dummy1_filter():
        value = trainer.state.epoch % 2
        print(f"trainer epoch {trainer.state.epoch} : "
              f"value={value}, "
              f"filter={value == 0}")
        return value == 0

    dummy1_on_the_fly = dummy1.attach_on_the_fly(evaluator, 'dummy1', dummy1_filter)

    @trainer.on(Events.EPOCH_COMPLETED)
    @dummy1_on_the_fly
    def _(_):
        # decoration on the fly
        # if filter ok, dummy1 is attached on the fly
        #if dummy1_filter():
        #    dummy1.attach(evaluator, 'dummy1')

        evaluator.run([0, 1, 2])

        # decoration on the fly
        # dummy1 is now detached
        #if dummy1_filter():
        #    dummy1.detach(evaluator)

    trainer.run([0, 1, 2, 3, 4, 5], max_epochs=5)

    print(f"dummy1 : reset={dummy1.call_count_reset} "
          f"compute={dummy1.call_count_compute} "
          f"update={dummy1.call_count_update}")
    print(f"dummy2 : reset={dummy2.call_count_reset} "
          f"compute={dummy2.call_count_compute} "
          f"update={dummy2.call_count_update}")

    assert(dummy1.call_count_reset == 3) # Remark: compute + 1 due to creation
    assert(dummy1.call_count_compute == 2)
    assert(dummy1.call_count_update == 6)

    assert(dummy2.call_count_reset == 6)
    assert(dummy2.call_count_compute == 5)
    assert(dummy2.call_count_update == 15)
```

I added following in `Metrics` 
```python
  def detach(self, engine: Engine) -> None:
      if engine.has_event_handler(self.completed, Events.EPOCH_COMPLETED):
          engine.remove_event_handler(self.completed, Events.EPOCH_COMPLETED)
      if engine.has_event_handler(self.started, Events.EPOCH_STARTED):
          engine.remove_event_handler(self.started, Events.EPOCH_STARTED)
      if engine.has_event_handler(self.iteration_completed, Events.ITERATION_COMPLETED):
          engine.remove_event_handler(self.iteration_completed, Events.ITERATION_COMPLETED)

  def attach_on_the_fly(self, engine: Engine, name: str, filter_fn: Callable) -> Callable:

      def decorator(f: Callable) -> Callable:

          def fun_(*args, **kwargs):
              if filter_fn():
                  self.attach(engine, name)
              f(*args, **kwargs)
              if filter_fn():
                  self.detach(engine)

          return fun_

      return decorator
```
@vfdev-5 any advice ? I didn't add engine ref in `Metrics` because metrics are implemented stateless atm.

**Comment by sdesrozis** [comment]
Alternative using `custom_event` :
```python
 def dummy1_filter(engine, event):
      value = engine.state.epoch % 2
      print(f"trainer epoch {trainer.state.epoch} : "
            f"value={value}, "
            f"filter={value == 0}")
      return value == 0

  @trainer.on(Events.EPOCH_COMPLETED(event_filter=dummy1_filter))
  def _(_):
      dummy1.attach(evaluator, 'dummy1')

  @trainer.on(Events.EPOCH_COMPLETED)
  def _(_):
      evaluator.run([0, 1, 2])

  @trainer.on(Events.EPOCH_COMPLETED(event_filter=dummy1_filter))
  def _(_):
      dummy1.detach(evaluator)
```

**Comment by vfdev-5** [comment]
@sdesrozis thanks for the snippets. I'll take a look in details a bit later. In general, I would say 
explicit calls will be more intuitive. Solution with `attach_on_the_fly` looks interesting but probably the API needs to be polished.

**Comment by sdesrozis** [comment]
@vfdev-5 Perfect. I'm testing these APIs in real code and finally `attach_on_the_fly` is not my preferred choice. The `event_filter` solution is great but decorators must be setted carefully. Don't know how far to go... My thought is we need hooks on events (per engine).

Btw, the minimal I see is to have a `detach` in metrics as I suggested previously.

**Comment by vfdev-5** [comment]
@sdesrozis I think it makes more sense to have `Metric` as other handlers without dependency on attached engine. So, yes, `detach` then should have engine as input argument. I think we can go on with this simple change of the API, adding `detach` only. In addition we can think to provide a helper method to setup a similar behaviour as `attach_on_the_fly` ?

**Comment by sdesrozis** [comment]
@vfdev-5 Provide a helper method is a good idea. I will keep trying to design a better solution because I’m not convinced by the API `attach_on_the_fly` that I proposed.

**Comment by sdesrozis** [comment]
@vfdev-5 PR for detach a metric ? (I suppose a test is required)

Consider the following snippet (`cm` standing for a `ConfusionMatrix` object) : 
```python
@trainer.on(Events.EPOCH_COMPLETED(event_filter=cm_filter))
def _(_):
  cm.attach(evaluator, 'cm')

@trainer.on(Events.EPOCH_COMPLETED)
def _(_):
  evaluator.run(vali_dataloader)

@trainer.on(Events.EPOCH_COMPLETED(event_filter=cm_filter))
def _(_):
  cm.detach(evaluator)
```
It works as expected but if a logger is attached on `evaluator`, it could warn. Indeed, tensor from `cm` can't be logged by tensorboard logger.

Notice that tag `all` is used in `setup_any_logging` for `OutputHandler`, so can't fix the warning with `common` tools. Is `metric_names` missing in the API of `setup_any_logging` ?

**Comment by vfdev-5** [comment]
> @vfdev-5 PR for detach a metric ? (I suppose a test is required)

@sdesrozis yes please with a test :)

> It works as expected but if a logger is attached on evaluator, it could warn. Indeed, tensor from cm can't be logged by tensorboard logger.

This part is a bit unclear for me.

**Comment by sdesrozis** [comment]
I admit that my post was very confusing, sorry! There is no link with the subject metric attach / detach... I think it's a new issue / question.

**Comment by vfdev-5** [comment]
@sdesrozis okay, I let you create new issue for that :)

**Comment by sdesrozis** [comment]
Proposition of helper `metric_on_the_fly` : 
```python
trainer = ...
train_evaluator = ...
vali_evaluator = ...

cm = ConfusionMatrix(num_classes=10, device="cuda")

def cm_filter(engine, event):
  return trainer.state.epoch == trainer.state.max_epochs

metric_on_the_fly(cm, train_evaluator, cm_filter)
metric_on_the_fly(cm, vali_evaluator, cm_filter)
```
with 
```python
def metric_on_the_fly(metric: Metric, metric_name: str, engine: Engine, event_filter: Callable) -> None:

  if not isinstance(metric, Metric):
    raise TypeError("Argument metric should be instance of Metric, but given {}".format(type(metric)))

  if not isinstance(engine, Engine):
    raise TypeError("Argument engine should be instance of Engine, but given {}".format(type(engine)))

  @engine.on(Events.STARTED(event_filter=event_filter))
  def _(_):
      metric.attach(engine, metric_name)

  @engine.on(Events.COMPLETED(event_filter=event_filter))
  def _(_):
      metric.detach(engine)
```
@vfdev-5 Thoughts?

In addition, this could be easily generalized to any attachable object and any event but I don't have use case yet.

**Comment by sdesrozis** [comment]
I keep playing to find something I like
```python
cm = ConfusionMatrix(num_classes=10, device=device)
cm_otf = OnTheFly(cm)
cm_otf.attach(tr_evaluator, "cm", last_epoch_filter(trainer))
cm_otf.attach(vl_evaluator, "cm", last_epoch_filter(trainer))
```
with the proxy `OnTheFly` to keep a coherent API
```python
class OnTheFly:
  def __init__(self, metric: Metric):
    self.metric = metric

  def attach(self, engine: Engine, name: str, event_filer: Callable):
      metric_on_the_fly(self.metric, name, engine, event_filer)
```
and `last_epoch`
```python
def last_epoch_filter(engine: Engine):
  def last_epoch(_, event):
      return engine.state.epoch == engine.state.max_epochs

  return last_epoch
```
@vfdev-5 Feel free to comment :)

**Comment by sdesrozis** [comment]
Concerning events, it would have been nice to decouple concepts `STARTED`, `COMPLETED` (bounding box) from the type of event `EPOCH`, `ITERATION` (counters), etc. 

For instance 
```python
Events.EPOCH_STARTED = Bounding.STARTED[Counter.EPOCH]

# can ask directly to a counter for create filters or reach info
max_epoch = Counter.EPOCH.last(trainer)
```

**Comment by sdesrozis** [comment]
I tried this API and I really enjoy 
```python
trainer = Engine(update)
tr_evaluator = Engine(inference)
vl_evaluator = Engine(inference)

with attach_metrics(trainer) as m:
    m["loss"] = RunningAverage(output_transform=lambda x: x)

with attach_metrics([tr_evaluator, vl_evaluator]) as m:

    m["nll"] = Loss(loss_fn=loss_fn)

    m["accuracy"] = Accuracy(device=device)

    precision = Precision(device=device)
    m["precision"] = precision
    m["avg_precision"] = precision.mean()

    recall = Recall(device=device)
    m["recall"] = recall
    m["avg_recall"] = recall.mean()

    f1 = Fbeta(beta=1, average=False, precision=precision, recall=recall, device=device)
    m["f1"] = f1
    m["avg_f1"] = f1.mean()

    # confusion matrix is expensive and computed on the fly
    cm = ConfusionMatrix(num_classes=10, device=device)
    m["cm"] = OnTheFly(cm, last_epoch_filter(trainer))
````
@vfdev-5 ?

**Comment by vfdev-5** [comment]
@sdesrozis are you talking about `OnTheFly` or context manager or both as "this" API ?

Personnally, I would rename `OnTheFly` by something else, the name is not obvious what it does.
About context managers to add metrics, well, I do not see the advantage of using it vs ordinary `m.attach(engine, "name")`. Could you please detail this, probably I'm missing something

**Comment by sdesrozis** [comment]
@vfdev-5 both :) 
I agree, `OnTheFly` is not suitable name. It applies `filter` to a metric. Don't know how to name this correctly atm.

For context manager, I see 2 advantages 
* it adds ... context :) don't write repetitive lines
* it adds feeling of scope (not real)
In my point of view, it improves api but it it's purely personal.

**Comment by vfdev-5** [comment]
@sdesrozis is this issue/question is still relevant ?

**Comment by sdesrozis** [comment]
@vfdev-5 I think we have discussed a lot around this. I close this issue.
